### PR TITLE
Add debug mode

### PR DESCRIPTION
**Pull request by denysoblohin-okta**
## Description:

Adds debug mode for SIW that allows inspecting console messages and fetch requests for environments where developer tools are not available (like Office 365).
Adds new file `okta-sign-in.debugger.min.js` to the build process output.
In order to enable debug mode it should be added to HTML before `okta-sign-in.*.js` AND initialised with `window.OktaSignInDebug.init()`
```
    <script src="%cdn%/okta-sign-in.debugger.min.js"></script>
    <script nonce="optional_nonce">window.OktaSignInDebug.init({ cspNonce: 'optional_nonce' });</script>
```
When loaded and initialised, it shows `Console` button at bottom-left corner of window. Click it to open popup with console logs. (See videos below.)
<s>SIW should be initialised with new option `debug: true` to allow catching console messages from SIW.</s>
Works on IE11 and CSP compliant.

### Test locally
Playground:
```
ENTRY=debugger yarn build:webpack-release
IE11_COMPAT_MODE=true yarn workspace v3 dev
```
(you also need to uncomment [this](https://github.com/okta/okta-signin-widget/pull/3546/files#diff-71c326312cbb9acee9247c5898197ab83ddd69d092d121e77450006bed7ed9c4R14))

Test app:
```
yarn build:webpack-release
TARGET=CROSS_BROWSER yarn start:test:app
```

Note: for testing on IE11 with ngrok need to add env vars `OKTA_SIW_HOST=0.0.0.0 DISABLE_CSP=1  `  , then use `ngrok http --host-header=rewrite 3000` or `npm install -g localtunnel && lt --port 3000 --local-host localhost`

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [x] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-683062](https://oktainc.atlassian.net/browse/OKTA-683062)

### Screenshot/Video:

https://github.com/okta/okta-signin-widget/assets/72614880/426fa7c7-0e5a-418c-ac3d-bed4c63adcd0

https://github.com/okta/okta-signin-widget/assets/72614880/a67b0ce8-4bb0-4ec0-9423-1172e11d47da

https://github.com/okta/okta-signin-widget/assets/72614880/4e99b04a-2058-4356-9c39-3318b69ae89b

(Note: error message on 14s is rendered by webpack dev server)

### Alternative:

POC using vConsole fork for IE11
https://github.com/denysoblohin-okta/vConsole/pull/1

https://github.com/okta/okta-signin-widget/assets/72614880/6c47902b-49c4-4719-99fb-eb686a500668

Found issues:
- IE11: Need to remove usage of [CSS vars](https://github.com/Tencent/vConsole/blob/dev/src/core/style/view.less#L26). Light/dark theming will not work without significant code changes.
- IE11: Need to fix usage of [`Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) for [`fetch`](https://github.com/Tencent/vConsole/blob/dev/src/network/fetch.proxy.ts) and `XMLHttpRequest`. [proxy-polyfill](https://github.com/GoogleChrome/proxy-polyfill) throws error if proxy handler has properties other than get/set/apply/construct.
- Text of switch button is hard-coded to "vConsole"
- Editing of cookie or local/session storage item value results in deletion of item
- IE11: Copy console message to clipboard doesn't work
- CSP issues

([PR](https://github.com/denysoblohin-okta/vConsole/pull/1) fixes only first 3 issues partially)

### CODE REVIEW

File: src/plugins/OktaPluginA11y.ts [hunk 1]
```diff
@@ -78,9 +79,11 @@ export const init = (widget: OktaSignInAPI): void => {
           const isVisible = input.type !== 'password';
           button.type = 'button';
           if (isVisible) {
+            // @ts-ignore
```
Comment (denysoblohin-okta) [inline src/plugins/OktaPluginA11y.ts:82]: Need to fix TS error here 
`error TS2339: Property 'ariaLabel' does not exist on type 'HTMLButtonElement'.` 
that came up after adding `"src/plugins/*"` to  [tsconfig.json](https://github.com/okta/okta-signin-widget/pull/3546/files#diff-b55cdbef4907b7045f32cc5360d48d262cca5f94062e353089f189f4460039e0)

File: rollup.release.config.js [hunk 2]
```diff
@@ -16,17 +16,7 @@ const config = buildRollupConfig(baseRollupConfig => {
...
-        'Logger.dir',
-        'Logger.time',
-        'Logger.timeEnd',
-        'Logger.group',
-        'Logger.groupEnd',
-        'Logger.assert',
-        'Logger.log',
-        'Logger.info',
-        'Logger.warn',
-        'Logger.deprecate'
```
Comment (glenfannin-okta) [inline rollup.release.config.js:29]: Would removing this here remove for all bundles? Can this be conditionally done somehow?

Comment (denysoblohin-okta) [reply rollup.release.config.js:29]: I think this should be for all bundles
@lesterchoi-okta What's your opinion?

File: package.json [hunk 3]
```diff
@@ -192,6 +192,7 @@
     "npm-run-all": "^4.1.5",
     "postcss": "^8.4.21",
     "postcss-loader": "^7.2.1",
+    "proxy-polyfill": "^0.3.2",
```
[the remark this answers is not in the document]
Comment (denysoblohin-okta) [reply package.json:195]: Should remove this. It's not used.

I tried to use `vConsole` library initially for a devtools simulation (use can see video in readme to have a n idea how it looks) and added this polyfill for IE11. But then rewritten from scratch.

File: src/plugins/OktaPluginA11y.ts (hunk 1 above)
Comment (denysoblohin-okta) [reply src/plugins/OktaPluginA11y.ts:83]: Looks like it's an issue with `typescript`'s DOM type defs.
Updating `typescript` from `~4.2` to `~4.4` resolves this issue.

Sadly can't easily align TS versions in root and `src/v3` (to `^4.9.5`) as there is an [issue](https://github.com/zerkalica/zerollup/issues/37) with `@zerollup/ts-transform-paths` plugin

File: src/util/Logger.js [hunk 4]
```diff
@@ -10,18 +10,25 @@
  * See the License for the specific language governing permissions and limitations under the License.
  */
 
+let debug = false;
+
 function log(level, args) {
   // Only log statements in development mode or if
   // throwing an error through console.error
-  if (DEBUG || level === 'error') {
+  if (debug || DEBUG || level === 'error') {
```
Comment (denysoblohin-okta) [inline src/util/Logger.js:18]: todo: use `window.OktaSignInDebug.isEnabled()` instead of local `debug`?
And remove `setDebug()`

**Comment by shuowu-okta** [comment]
Great to see debugging mode works in SIW! 

Some suggestions / questions:
1. expose devMode as a SIW config instead of a global (window) variable
2. `OktaSignInDebug` as a monorepo workspace if it's fully self contained, so it can be easily imported, and avoid relative path issues.
3. maybe consider pass it to widget as option instead of bundle it into widget. Then it can be shared globally via React context / Setting model
4. what fetch.ts is for in debugger plugin?

File: src/util/Logger.js (hunk 4 above)
Comment (denysoblohin-okta) [reply src/util/Logger.js]: Or rename `debug` options in SIW to `verbose` which is more appropriate name

File: src/plugins/OktaPluginDebugger/utils.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    String(d.getHours()).padStart(2, '0'),
+    ':',
+    String(d.getMinutes()).padStart(2, '0'),
+    ':',
+    String(d.getSeconds()).padStart(2, '0'),
+    ' .',
+    String(d.getMilliseconds()).padStart(3, '0'),
+  ];
+  return parts.join('');
+};
```
Comment (lesterchoi-okta) [inline src/plugins/OktaPluginDebugger/utils.ts:113]:
```suggestion
export const getCurrentTimeStamp = () => {
  return new Date().toISOString();
};
```

File: rollup.release.config.js (hunk 2 above)
Comment (lesterchoi-okta) [reply rollup.release.config.js:29]: @glenfannin-okta what would the condition be?

File: src/plugins/OktaPluginDebugger/messages.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    lines.push('Meta: ' + serializeValue(fetchMeta));
+    if (err) {
+      lines.push(...serializeArgs([err]));
+    }
+  } else {
+    lines.push(...serializeArgs(args));
+  }
+  lines[0] = `#${String(no).padStart(4, ' ')}. ${time} [${method}]  ` + lines[0];
+  return lines.join('\n');
```
Comment (lesterchoi-okta) [inline src/plugins/OktaPluginDebugger/messages.ts:96]: @denysoblohin-okta OK for now to unblock urgent customer issues, but we should aim for more structured log messages (see [this article about log format best practices](https://betterstack.com/community/guides/logging/log-formatting/#3-structured-logging)).
We'll probably need a way to redact sensitive fields similar to Sentry.

**Review by lesterchoi-okta — APPROVED** [review]
Good for now to unblock issues, but let's fast-follow on the message format and redaction.

**Review by lesterchoi-okta — CHANGES_REQUESTED** [review]
Sorry for the premature "approval", please change the timestamp format to use `Date.prototype.toISOString()`

File: playground/main.ts [hunk 7]
```diff
@@ -44,6 +44,10 @@ function getWidgetInstance() {
 }
 
 function createWidgetInstance(options: WidgetOptions = {}) {
+  options = {
+    ...options,
+    debug: true,
```
[the remark this answers is not in the document]
Comment (denysoblohin-okta) [reply playground/main.ts]: I've removed `debug` from options. 
https://github.com/okta/okta-signin-widget/pull/3546/files#diff-bf3aa7d3e0f122607ccc9a328eda2b2409fbb00d66959e285e0a5aa222cd5909R17

File: src/plugins/OktaPluginA11y.ts [hunk 8]
```diff
@@ -9,7 +10,7 @@ import { EventContext } from '../types/events';
  * @param widget instance of OktaSignIn
  * @param options options for a11y plugin
  */
-export const init = (widget: OktaSignInAPI): void => {
+export const init = (widget: OktaSignInAPI & { options: WidgetOptions }): void => {
```
[the remark this answers is not in the document]
Comment (denysoblohin-okta) [reply src/plugins/OktaPluginA11y.ts:13]: Yes. `Property 'options' does not exist on type 'OktaSignInAPI'`

File: rollup.release.config.js (hunk 2 above)
Comment (glenfannin-okta) [reply rollup.release.config.js:29]: Actually, after looking through it, I think this is sufficient: https://github.com/okta/okta-signin-widget/pull/3546/files#diff-bf3aa7d3e0f122607ccc9a328eda2b2409fbb00d66959e285e0a5aa222cd5909R17

Basically, I just wanted to be sure we arent exposting these commands when we didnt intend to in production environments. I think the above changes achieves that.

**Review by lesterchoi-okta — APPROVED** [review]

**Review by lesterchoi-okta — COMMENTED** [review]
LGTM @shuowu-okta

**Pull request merged**